File: packages/control-client/rust/lib/request.rs
//! Prepared checked operations. No SDK resource convergence policy lives here.

use microsandbox_protocol::{
    control::{
        BranchCreate, BranchResult, Capabilities, CheckpointCreate, CheckpointResult, ControlError,
        CpuState, CpuTarget, DiskCheckpointCreate, DiskCheckpointState, DiskCompact, Empty,
        MemoryState, MemoryTarget, Pause, PauseState, RootDiskGrow, RootDiskState,
        RuntimeCapabilities, SecretChange, SecretsResult,
    },
    wire,
};
use microsandbox_protocol_client::{EncodedMessage, Message, Request};
use microsandbox_utils::size::Mebibytes;
use serde::{Serialize, de::DeserializeOwned};

use crate::{
    CompatibleControlRequest, ControlClientError, ControlClientResult, ControlProtocol, JsonReply,
};
use zeroize::Zeroizing;

//--------------------------------------------------------------------------------------------------
// Types
//--------------------------------------------------------------------------------------------------

/// Query available host operations.
#[derive(Debug, Clone, Copy, Default)]
pub struct GetCapabilities;
/// Query the complete generation-two runtime facility inventory.
#[derive(Debug, Clone, Copy, Default)]
pub struct GetRuntimeCapabilities;
/// Read accepted and observed memory quantities.
#[derive(Debug, Clone, Copy, Default)]
pub struct GetMemoryState;
/// Set a memory target without waiting for guest convergence.
#[derive(Debug, Clone, Copy)]
pub struct SetMemoryTarget {
    /// Full-width wire quantity; direct construction avoids SDK input narrowing.
    pub total_mib: u64,
}
/// Read CPU capacity, target, observation, and enforcement.
#[derive(Debug, Clone, Copy, Default)]
pub struct GetCpuState;
/// Set a CPU target without waiting for guest convergence.
#[derive(Debug, Clone, Copy)]
pub struct SetCpuTarget {
    /// Requested online CPUs.
    pub online: u32,
}
/// Apply ordered secret changes, preserving partial completion in the result.
#[derive(Debug, Clone)]
pub struct UpdateSecrets {
    /// Entries execute sequentially, stopping at the first operation failure.
    pub changes: Vec<SecretChange>,
}
/// Create one full checkpoint.
#[derive(Debug, Clone)]
pub struct CreateCheckpoint(pub CheckpointCreate);
/// Create one disk-only checkpoint.
#[derive(Debug, Clone)]
pub struct CreateDiskCheckpoint(pub DiskCheckpointCreate);
/// Create one direct local branch without descriptor transfer.
#[derive(Debug, Clone)]
pub struct CreateBranch(pub BranchCreate);
/// Pause the runtime, optionally requiring guest writeback.
#[derive(Debug, Clone, Copy, Default)]
pub struct PauseRuntime(pub Pause);
/// Resume a resident pause.
#[derive(Debug, Clone, Copy, Default)]
pub struct ResumeRuntime;
/// Inspect resident pause state.
#[derive(Debug, Clone, Copy, Default)]
pub struct GetPauseState;
/// Grow the owned root disk and filesystem.
#[derive(Debug, Clone, Copy)]
pub struct GrowRootDisk(pub RootDiskGrow);
/// Compact selected owned disk chains.
#[derive(Debug, Clone)]
pub struct CompactDisks(pub DiskCompact);

//--------------------------------------------------------------------------------------------------
// Methods
//--------------------------------------------------------------------------------------------------

impl SetMemoryTarget {
    /// Accept the SDK's existing integer/MiB helpers and their conversion rules.
    ///
    /// `SetMemoryTarget::new(2048.mib())` uses the shared `SizeExt` owner. This
    /// does not change its existing sub-MiB truncation or overflow semantics.
    /// For full-width MiB input construct the public `total_mib` field directly.
    pub fn new(size: impl Into<Mebibytes>) -> Self {
        Self {
            total_mib: u64::from(size.into().as_u32()),
        }
    }
}

impl SetCpuTarget {
    /// Prepare an online CPU target without performing I/O.
    pub fn new(online: u32) -> Self {
        Self { online }
    }
}

impl UpdateSecrets {
    /// Prepare a caller-ordered batch without performing I/O.
    pub fn new(changes: Vec<SecretChange>) -> Self {
        Self { changes }
    }
}

//--------------------------------------------------------------------------------------------------
// Trait Implementations
//--------------------------------------------------------------------------------------------------

impl Request<ControlProtocol> for GetCapabilities {
    type Response = Capabilities;
    type Error = ControlClientError;
    fn message(&self) -> ControlClientResult<EncodedMessage> {
        prepared("control.capabilities", &Empty {})
    }
    fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
        checked(response, 1, "control.capabilities.result")
    }
}

impl Request<ControlProtocol> for GetRuntimeCapabilities {
    type Response = RuntimeCapabilities;
    type Error = ControlClientError;
    fn message(&self) -> ControlClientResult<EncodedMessage> {
        prepared("control.capabilities", &Empty {})
    }
    fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
        checked(response, 2, "control.capabilities.result")
    }
}

impl Request<ControlProtocol> for GetMemoryState {
    type Response = MemoryState;
    type Error = ControlClientError;
    fn message(&self) -> ControlClientResult<EncodedMessage> {
        prepared("control.memory.state", &Empty {})
    }
    fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
        checked(response, 1, "control.memory.state")
    }
}

impl Request<ControlProtocol> for SetMemoryTarget {
    type Response = MemoryState;
    type Error = ControlClientError;
    fn message(&self) -> ControlClientResult<EncodedMessage> {
        prepared(
            "control.memory.target",
            &MemoryTarget {
                total_mib: self.total_mib,
            },
        )
    }
    fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
        checked(response, 1, "control.memory.state")
    }
}

impl Request<ControlProtocol> for GetCpuState {
    type Response = CpuState;
    type Error = ControlClientError;
    fn message(&self) -> ControlClientResult<EncodedMessage> {
        prepared("control.cpu.state", &Empty {})
    }
    fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
        checked(response, 1, "control.cpu.state")
    }
}

impl Request<ControlProtocol> for SetCpuTarget {
    type Response = CpuState;
    type Error = ControlClientError;
    fn message(&self) -> ControlClientResult<EncodedMessage> {
        prepared(
            "control.cpu.target",
            &CpuTarget {
                online: self.online,
            },
        )
    }
    fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
        checked(response, 1, "control.cpu.state")
    }
}

impl Request<ControlProtocol> for UpdateSecrets {
    type Response = SecretsResult;
    type Error = ControlClientError;
    fn message(&self) -> ControlClientResult<EncodedMessage> {
        #[derive(Serialize)]
        struct Payload<'a> {
            changes: &'a [SecretChange],
        }
        prepared(
            "control.secrets.update",
            &Payload {
                changes: &self.changes,
            },
        )
    }
    fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
        checked_with(response, 1, "control.secrets.result", |bytes| {
            let result = SecretsResult::decode(bytes)?;
            // Completion cannot include entries the caller never sent. Preserve
            // partial failure as an ordinary typed result, not a rollback claim.
            let valid = match &result {
                SecretsResult::Complete { applied_count } => {
                    *applied_count as usize == self.changes.len()
                }
                SecretsResult::Failed { failed_index, .. } => {
                    (*failed_index as usize) < self.changes.len()
                }
            };
            if !valid {
                return Err(wire::WireError::InvalidRecord);
            }
            Ok(result)
        })
    }
}

macro_rules! v2_request {
    ($request:ty, $response:ty, $request_name:literal, $response_name:literal) => {
        impl Request<ControlProtocol> for $request {
            type Response = $response;
            type Error = ControlClientError;

            fn message(&self) -> ControlClientResult<EncodedMessage> {
                prepared($request_name, &self.0)
            }

            fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
                checked(response, 2, $response_name)
            }
        }
    };
}

v2_request!(
    CreateCheckpoint,
    CheckpointResult,
    "control.checkpoint.create",
    "control.checkpoint.result"
);
v2_request!(
    CreateDiskCheckpoint,
    DiskCheckpointState,
    "control.disk.checkpoint.create",
    "control.disk.checkpoint.result"
);
v2_request!(
    CreateBranch,
    BranchResult,
    "control.branch.create",
    "control.branch.result"
);
v2_request!(
    PauseRuntime,
    PauseState,
    "control.pause",
    "control.pause.state"
);
v2_request!(
    GrowRootDisk,
    RootDiskState,
    "control.root-disk.grow",
    "control.root-disk.state"
);
v2_request!(
    CompactDisks,
    microsandbox_protocol::control::DiskCompactionResult,
    "control.disk.compact",
    "control.disk.compact.result"
);

impl Request<ControlProtocol> for ResumeRuntime {
    type Response = PauseState;
    type Error = ControlClientError;

    fn message(&self) -> ControlClientResult<EncodedMessage> {
        prepared("control.resume", &Empty {})
    }

    fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
        checked(response, 2, "control.pause.state")
    }
}

impl Request<ControlProtocol> for GetPauseState {
    type Response = PauseState;
    type Error = ControlClientError;

    fn message(&self) -> ControlClientResult<EncodedMessage> {
        prepared("control.pause.state", &Empty {})
    }

    fn decode(&self, response: Message) -> ControlClientResult<Self::Response> {
        checked(response, 2, "control.pause.state")
    }
}

macro_rules! compatible_v2 {
    ($request:ty, $json:expr, $decode:expr) => {
        impl CompatibleControlRequest for $request {
            fn min_generation(&self) -> u8 {
                2
            }

            fn compatibility_json_bytes(&self) -> ControlClientResult<Zeroizing<Vec<u8>>> {
                let value = ($json)(self);
                Ok(Zeroizing::new(serde_json::to_vec(&value).map_err(
                    |_| {
                        microsandbox_protocol_client::ClientError::new(
                            microsandbox_protocol_client::ErrorKind::Encode,
                        )
                    },
                )?))
            }

            fn decode_compatibility_json(
                &self,
                reply: JsonReply,
            ) -> ControlClientResult<Self::Response> {
                ($decode)(reply)
            }
        }
    };
}

compatible_v2!(
    GetRuntimeCapabilities,
    |_: &GetRuntimeCapabilities| serde_json::json!({
        "op": "capabilities",
    }),
    |reply| decode_json_field(reply, "capabilities")
);

compatible_v2!(
    CreateCheckpoint,
    |request: &CreateCheckpoint| serde_json::json!({
        "op": "checkpoint_create",
        "guest_flush": request.0.guest_flush,
        "record_integrity": request.0.record_integrity,
        "checkpoint_id": request.0.checkpoint_id,
        "intent": request.0.intent,
    }),
    decode_checkpoint_json
);
compatible_v2!(
    CreateDiskCheckpoint,
    |request: &CreateDiskCheckpoint| serde_json::json!({
        "op": "disk_checkpoint_create",
        "guest_flush": request.0.guest_flush,
        "checkpoint_id": request.0.checkpoint_id,
    }),
    |reply| decode_json_field(reply, "disk_checkpoint")
);
compatible_v2!(
    CreateBranch,
    |request: &CreateBranch| serde_json::json!({
        "op": "branch_create",
        "guest_flush": request.0.guest_flush,
        "record_integrity": request.0.record_integrity,
        "branch_id": request.0.branch_id,
        "child_name": request.0.child_name,
        "memory_cache_dir": request.0.memory_cache_dir,
    }),
    decode_branch_json
);
compatible_v2!(
    PauseRuntime,
    |request: &PauseRuntime| match request.0.guest_flush {
        Some(policy) => serde_json::json!({"op": "pause_with_guest_flush", "guest_flush": policy}),
        None => serde_json::json!({"op": "pause"}),
    },
    |reply| decode_json_field(reply, "pause")
);
compatible_v2!(
    GrowRootDisk,
    |request: &GrowRootDisk| serde_json::json!({
        "op": "root_disk_grow", "size_bytes": request.0.size_bytes,
    }),
    |reply| decode_json_field(reply, "root_disk")
);
compatible_v2!(
    CompactDisks,
    |request: &CompactDisks| serde_json::json!({
        "op": "disk_compact",
        "target": request.0.target,
        "layers": request.0.layers,
        "dry_run": request.0.dry_run,
    }),
    |reply| decode_json_field(reply, "compaction")
);

impl CompatibleControlRequest for ResumeRuntime {
    fn min_generation(&self) -> u8 {
        2
    }
    fn compatibility_json_bytes(&self) -> ControlClientResult<Zeroizing<Vec<u8>>> {
        json_operation("resume")
    }
    fn decode_compatibility_json(&self, reply: JsonReply) -> ControlClientResult<Self::Response> {
        decode_json_field(reply, "pause")
    }
}

impl CompatibleControlRequest for GetPauseState {
    fn min_generation(&self) -> u8 {
        2
    }
    fn compatibility_json_bytes(&self) -> ControlClientResult<Zeroizing<Vec<u8>>> {
        json_operation("pause_state")
    }
    fn decode_compatibility_json(&self, reply: JsonReply) -> ControlClientResult<Self::Response> {
        decode_json_field(reply, "pause")
    }
}

//--------------------------------------------------------------------------------------------------
// Functions
//--------------------------------------------------------------------------------------------------

fn prepared(name: &str, payload: &impl Serialize) -> ControlClientResult<EncodedMessage> {
    Ok(EncodedMessage::new(name, wire::encode(payload)?))
}

fn checked<T: DeserializeOwned>(
    response: Message,
    minimum: u8,
    expected: &str,
) -> ControlClientResult<T> {
    checked_with(response, minimum, expected, wire::decode_record)
}

fn checked_with<T>(
    response: Message,
    minimum: u8,
    expected: &str,
    decode: impl FnOnce(&[u8]) -> Result<T, wire::WireError>,
) -> ControlClientResult<T> {
    if !(minimum..=microsandbox_protocol::control::CONTROL_GENERATION).contains(&response.v)
        || response.id == 0
        || response.flags != 1
    {
        return Err(ControlClientError::InvalidResponse {
            response: Box::new(response),
        });
    }
    if response.t == "control.error" {
        let Ok(error) = wire::decode_record::<ControlError>(&response.p) else {
            return Err(ControlClientError::InvalidResponse {
                response: Box::new(response),
            });
        };
        return Err(ControlClientError::Peer {
            error,
            response: Box::new(response),
        });
    }
    if response.t != expected {
        return Err(ControlClientError::InvalidResponse {
            response: Box::new(response),
        });
    }
    decode(&response.p).map_err(|_| ControlClientError::InvalidResponse {
        response: Box::new(response),
    })
}

fn json_operation(operation: &str) -> ControlClientResult<Zeroizing<Vec<u8>>> {
    Ok(Zeroizing::new(
        serde_json::to_vec(&serde_json::json!({"op": operation})).map_err(|_| {
            microsandbox_protocol_client::ClientError::new(
                microsandbox_protocol_client::ErrorKind::Encode,
            )
        })?,
    ))
}

fn decode_json_field<T: DeserializeOwned>(reply: JsonReply, field: &str) -> ControlClientResult<T> {
    let decoded = serde_json::from_slice::<serde_json::Value>(reply.raw())
        .ok()
        .and_then(|value| value.get(field).cloned())
        .and_then(|value| serde_json::from_value(value).ok());
    reply.checked(|_| decoded)
}

fn decode_branch_json(reply: JsonReply) -> ControlClientResult<BranchResult> {
    let path = serde_json::from_slice::<serde_json::Value>(reply.raw())
        .ok()
        .and_then(|value| value.get("branch").cloned())
        .and_then(|value| serde_json::from_value(value).ok());
    reply.checked(|_| path.map(|path| BranchResult { path }))
}

fn decode_checkpoint_json(reply: JsonReply) -> ControlClientResult<CheckpointResult> {
    #[derive(serde::Deserialize)]
    struct Response {
        ok: bool,
        #[serde(default)]
        error: Option<String>,
        #[serde(default)]
        checkpoint: Option<microsandbox_protocol::control::CheckpointState>,
    }

    let decoded = serde_json::from_slice::<Response>(reply.raw()).ok();
    if let Some(response) = decoded
        && let Some(checkpoint) = response.checkpoint
    {
        return Ok(CheckpointResult {
            checkpoint: Some(checkpoint),
            recovery_error: (!response.ok).then_some(response.error).flatten(),
        });
    }
    reply.checked(|_| None)
}

File: packages/control-client/typescript/src/request.ts
import { encodedMessage, encodeRecord, WireError, type EncodedMessage, type InboundFrame } from "@microsandbox/protocol-client";
import type { Mebibytes } from "@microsandbox/types/size";
import { ControlClientError } from "./error.js";
import {
  jsonBranch, jsonCapabilities, jsonCheckpoint, jsonCpu, jsonDiskCheckpoint,
  jsonDiskCompaction, jsonMemory, jsonPause, jsonRootDisk, jsonRuntimeCapabilities,
  type JsonReply,
} from "./json-reply.js";
import {
  snapshotChanges, type CheckedControlRequest, type CompatibleControlRequest,
  type ExtendedLegacyControlRequest, type LegacyControlRequest,
} from "./legacy-request.js";
import {
  CONTROL_GENERATION, checkedUint, decodeBranchResult, decodeCapabilities, decodeCheckpointResult,
  decodeControlError, decodeCpuState, decodeDiskCheckpointState, decodeDiskCompactionResult,
  decodeMemoryState, decodePauseState, decodeRootDiskState, decodeRuntimeCapabilities,
  decodeSecretsResult, type BranchCreate, type BranchResult, type Capabilities,
  type CheckpointCreate, type CheckpointResult, type CpuState, type DiskCheckpointCreate,
  type DiskCheckpointState, type DiskCompact, type DiskCompactionResult, type MemoryState,
  type Pause, type PauseState, type RootDiskGrow, type RootDiskState, type RuntimeCapabilities,
  type SecretChange, type SecretsResult,
} from "./records.js";

function checked<T>(frame: InboundFrame, name: string, decode: (bytes: Uint8Array) => T, minimum = 1): T {
  const invalid = () => new ControlClientError("invalid_response", frame);
  if (frame.protocolVersion < minimum || frame.protocolVersion > CONTROL_GENERATION || frame.id === 0 || frame.flags !== 1) throw invalid();
  if (frame.type === "control.error") {
    let error;
    try { error = decodeControlError(frame.payload); } catch { throw invalid(); }
    throw new ControlClientError("peer", frame, error);
  }
  if (frame.type !== name) throw invalid();
  try { return decode(frame.payload); } catch { throw invalid(); }
}

/** Optional checked helpers only prepare records and decode replies; no SDK policy. */
export class GetCapabilities implements CheckedControlRequest<Capabilities> {
  message(): EncodedMessage { return encodedMessage("control.capabilities", encodeRecord({})); }
  decode(frame: InboundFrame): Capabilities { return checked(frame, "control.capabilities.result", decodeCapabilities); }
  jsonRequest(): LegacyControlRequest { return { op: "capabilities" }; }
  decodeJson(reply: JsonReply): Capabilities { return reply.checked(value => jsonCapabilities(value.get("capabilities"))); }
}
export class GetRuntimeCapabilities implements CompatibleControlRequest<RuntimeCapabilities> {
  readonly minGeneration = 2;
  message(): EncodedMessage { return encodedMessage("control.capabilities", encodeRecord({})); }
  decode(frame: InboundFrame): RuntimeCapabilities {
    return checked(frame, "control.capabilities.result", decodeRuntimeCapabilities, 2);
  }
  jsonRequest(): LegacyControlRequest { return { op: "capabilities" }; }
  decodeJson(reply: JsonReply): RuntimeCapabilities {
    return reply.checked(value => jsonRuntimeCapabilities(value.get("capabilities")));
  }
}
export class GetMemoryState implements CheckedControlRequest<MemoryState> {
  message(): EncodedMessage { return encodedMessage("control.memory.state", encodeRecord({})); }
  decode(frame: InboundFrame): MemoryState { return checked(frame, "control.memory.state", decodeMemoryState); }
  jsonRequest(): LegacyControlRequest { return { op: "memory_state" }; }
  decodeJson(reply: JsonReply): MemoryState { return reply.checked(value => jsonMemory(value.get("memory"))); }
}
export class SetMemoryTarget implements CheckedControlRequest<MemoryState> {
  readonly total_mib: bigint;
  /** Reuse SDK units; bigint also accepts a full-width wire quantity directly. */
  constructor(size: Mebibytes | bigint) { this.total_mib = checkedUint(size, 64); }
  message(): EncodedMessage { return encodedMessage("control.memory.target", encodeRecord({ total_mib: this.total_mib })); }
  decode(frame: InboundFrame): MemoryState { return checked(frame, "control.memory.state", decodeMemoryState); }
  jsonRequest(): LegacyControlRequest { return { op: "memory_target", total_mib: this.total_mib }; }
  decodeJson(reply: JsonReply): MemoryState { return new GetMemoryState().decodeJson(reply); }
}
export class GetCpuState implements CheckedControlRequest<CpuState> {
  message(): EncodedMessage { return encodedMessage("control.cpu.state", encodeRecord({})); }
  decode(frame: InboundFrame): CpuState { return checked(frame, "control.cpu.state", decodeCpuState); }
  jsonRequest(): LegacyControlRequest { return { op: "cpu_state" }; }
  decodeJson(reply: JsonReply): CpuState { return reply.checked(value => jsonCpu(value.get("cpu"))); }
}
export class SetCpuTarget implements CheckedControlRequest<CpuState> {
  readonly online: number;
  constructor(online: number) { this.online = Number(checkedUint(online, 32)); }
  message(): EncodedMessage { return encodedMessage("control.cpu.target", encodeRecord({ online: this.online })); }
  decode(frame: InboundFrame): CpuState { return checked(frame, "control.cpu.state", decodeCpuState); }
  jsonRequest(): LegacyControlRequest { return { op: "cpu_target", online: this.online }; }
  decodeJson(reply: JsonReply): CpuState { return new GetCpuState().decodeJson(reply); }
}
export class UpdateSecrets implements CheckedControlRequest<SecretsResult> {
  readonly #changes: readonly SecretChange[];
  constructor(changes: readonly SecretChange[]) {
    // Snapshot the prepared batch so caller mutation cannot change the sent
    // entries or invalidate interpretation of partial progress after admission.
    this.#changes = snapshotChanges(changes);
  }
  message(): EncodedMessage { return encodedMessage("control.secrets.update", encodeRecord({ changes: this.#changes })); }
  jsonRequest(): LegacyControlRequest { return { op: "secrets_update", changes: snapshotChanges(this.#changes) }; }
  decodeJson(reply: JsonReply): SecretsResult {
    return reply.checked(() => ({ outcome: "complete", applied_count: this.#changes.length }));
  }
  decode(frame: InboundFrame): SecretsResult {
    return checked(frame, "control.secrets.result", bytes => {
      const result = decodeSecretsResult(bytes);
      if (result.outcome === "complete" ? result.applied_count !== this.#changes.length : result.failed_index >= this.#changes.length) throw new WireError("invalid_record");
      return result;
    });
  }
}

abstract class GenerationTwoRequest<T> implements CompatibleControlRequest<T> {
  readonly minGeneration = 2;
  abstract message(): EncodedMessage;
  abstract decode(frame: InboundFrame): T;
  abstract jsonRequest(): ExtendedLegacyControlRequest;
  abstract decodeJson(reply: JsonReply): T;
}

export class CreateCheckpoint extends GenerationTwoRequest<CheckpointResult> {
  readonly value: CheckpointCreate;
  constructor(value: CheckpointCreate) {
    super();
    const intent = value.intent;
    if (intent !== "full_snapshot" && intent !== "park" && intent !== "transparent_transfer") throw new WireError("invalid_record");
    this.value = {
      ...(value.guest_flush === undefined ? {} : { guest_flush: guestFlush(value.guest_flush) }),
      record_integrity: bool(value.record_integrity), checkpoint_id: text(value.checkpoint_id), intent,
    };
  }
  message(): EncodedMessage { return encodedMessage("control.checkpoint.create", encodeRecord(compact(this.value))); }
  decode(frame: InboundFrame): CheckpointResult {
    return checked(frame, "control.checkpoint.result", decodeCheckpointResult, 2);
  }
  jsonRequest(): ExtendedLegacyControlRequest { return { op: "checkpoint_create", ...this.value }; }
  decodeJson(reply: JsonReply): CheckpointResult {
    let checkpoint: CheckpointResult["checkpoint"];
    try { checkpoint = reply.value.has("checkpoint") ? jsonCheckpoint(reply.value.get("checkpoint")) : undefined; }
    catch { throw new ControlClientError("invalid_json_response", reply); }
    const ok = reply.value.get("ok");
    if (checkpoint && (ok === true || ok === false)) {
      const error = reply.value.get("error");
      if (error !== undefined && error !== null && typeof error !== "string") throw new ControlClientError("invalid_json_response", reply);
      return { checkpoint, ...(ok === false ? { recovery_error: typeof error === "string" ? error : "source recovery failed without a runtime diagnostic" } : {}) };
    }
    return reply.checked(() => { throw new ControlClientError("invalid_json_response", reply); });
  }
}

export class CreateDiskCheckpoint extends GenerationTwoRequest<DiskCheckpointState> {
  readonly value: DiskCheckpointCreate;
  constructor(value: DiskCheckpointCreate) {
    super();
    this.value = {
      ...(value.guest_flush === undefined ? {} : { guest_flush: guestFlush(value.guest_flush) }),
      checkpoint_id: text(value.checkpoint_id),
    };
  }
  message(): EncodedMessage { return encodedMessage("control.disk.checkpoint.create", encodeRecord(compact(this.value))); }
  decode(frame: InboundFrame): DiskCheckpointState {
    return checked(
      frame,
      "control.disk.checkpoint.result",
      bytes => decodeDiskCheckpointState(bytes, frame.decodePayload()),
      2,
    );
  }
  jsonRequest(): ExtendedLegacyControlRequest { return { op: "disk_checkpoint_create", ...this.value }; }
  decodeJson(reply: JsonReply): DiskCheckpointState {
    return reply.checked(value => jsonDiskCheckpoint(value.get("disk_checkpoint")));
  }
}

export class CreateBranch extends GenerationTwoRequest<BranchResult> {
  readonly value: BranchCreate;
  constructor(value: BranchCreate) {
    super();
    this.value = {
      ...(value.guest_flush === undefined ? {} : { guest_flush: guestFlush(value.guest_flush) }),
      record_integrity: bool(value.record_integrity), branch_id: text(value.branch_id),
      child_name: text(value.child_name), memory_cache_dir: text(value.memory_cache_dir),
    };
  }
  message(): EncodedMessage { return encodedMessage("control.branch.create", encodeRecord(compact(this.value))); }
  decode(frame: InboundFrame): BranchResult { return checked(frame, "control.branch.result", decodeBranchResult, 2); }
  jsonRequest(): ExtendedLegacyControlRequest { return { op: "branch_create", ...this.value }; }
  decodeJson(reply: JsonReply): BranchResult { return reply.checked(value => jsonBranch(value.get("branch"))); }
}

export class PauseRuntime extends GenerationTwoRequest<PauseState> {
  readonly value: Pause;
  constructor(value: Pause = {}) {
    super();
    this.value = value.guest_flush === undefined ? {} : { guest_flush: guestFlush(value.guest_flush) };
  }
  message(): EncodedMessage { return encodedMessage("control.pause", encodeRecord(compact(this.value))); }
  decode(frame: InboundFrame): PauseState { return checked(frame, "control.pause.state", decodePauseState, 2); }
  jsonRequest(): ExtendedLegacyControlRequest {
    return this.value.guest_flush === undefined
      ? { op: "pause" }
      : { op: "pause_with_guest_flush", guest_flush: this.value.guest_flush };
  }
  decodeJson(reply: JsonReply): PauseState { return reply.checked(value => jsonPause(value.get("pause"))); }
}

export class ResumeRuntime extends GenerationTwoRequest<PauseState> {
  message(): EncodedMessage { return encodedMessage("control.resume", encodeRecord({})); }
  decode(frame: InboundFrame): PauseState { return checked(frame, "control.pause.state", decodePauseState, 2); }
  jsonRequest(): ExtendedLegacyControlRequest { return { op: "resume" }; }
  decodeJson(reply: JsonReply): PauseState { return reply.checked(value => jsonPause(value.get("pause"))); }
}

export class GetPauseState extends GenerationTwoRequest<PauseState> {
  message(): EncodedMessage { return encodedMessage("control.pause.state", encodeRecord({})); }
  decode(frame: InboundFrame): PauseState { return checked(frame, "control.pause.state", decodePauseState, 2); }
  jsonRequest(): ExtendedLegacyControlRequest { return { op: "pause_state" }; }
  decodeJson(reply: JsonReply): PauseState { return reply.checked(value => jsonPause(value.get("pause"))); }
}

export class GrowRootDisk extends GenerationTwoRequest<RootDiskState> {
  readonly value: RootDiskGrow;
  constructor(sizeBytes: number | bigint) { super(); this.value = { size_bytes: checkedUint(sizeBytes, 64) }; }
  message(): EncodedMessage { return encodedMessage("control.root-disk.grow", encodeRecord(this.value)); }
  decode(frame: InboundFrame): RootDiskState { return checked(frame, "control.root-disk.state", decodeRootDiskState, 2); }
  jsonRequest(): ExtendedLegacyControlRequest { return { op: "root_disk_grow", ...this.value }; }
  decodeJson(reply: JsonReply): RootDiskState { return reply.checked(value => jsonRootDisk(value.get("root_disk"))); }
}

export class CompactDisks extends GenerationTwoRequest<DiskCompactionResult> {
  readonly value: DiskCompact;
  constructor(value: Omit<DiskCompact, "layers"> & { layers?: number | bigint }) {
    super();
    const target = value.target;
    const checkedTarget = target.kind === "all" || target.kind === "root"
      ? { kind: target.kind }
      : target.kind === "disk"
        ? { kind: "disk" as const, guest_path: text(target.guest_path) }
        : (() => { throw new WireError("invalid_record"); })();
    this.value = {
      target: checkedTarget, dry_run: bool(value.dry_run),
      ...(value.layers === undefined ? {} : { layers: checkedUint(value.layers, 64) }),
    };
  }
  message(): EncodedMessage { return encodedMessage("control.disk.compact", encodeRecord(compact(this.value))); }
  decode(frame: InboundFrame): DiskCompactionResult {
    return checked(frame, "control.disk.compact.result", decodeDiskCompactionResult, 2);
  }
  jsonRequest(): ExtendedLegacyControlRequest { return { op: "disk_compact", ...this.value }; }
  decodeJson(reply: JsonReply): DiskCompactionResult {
    return reply.checked(value => jsonDiskCompaction(value.get("compaction")));
  }
}

function compact<T extends object>(value: T): T {
  return Object.fromEntries(Object.entries(value).filter(([, field]) => field !== undefined)) as T;
}

function text(value: unknown): string {
  if (typeof value !== "string" || /[\uD800-\uDFFF]/u.test(value)) throw new WireError("invalid_record");
  return value;
}

function bool(value: unknown): boolean {
  if (typeof value !== "boolean") throw new WireError("invalid_record");
  return value;
}

function guestFlush(value: unknown): "auto" | "required" | "skip" {
  if (value !== "auto" && value !== "required" && value !== "skip") throw new WireError("invalid_record");
  return value;
}
